src/bundle/Security/TwoFactor/Provider/TwoFactorProviderInterface.php

```diff
@@ -13,6 +13,11 @@ interface TwoFactorProviderInterface
      */
     public function beginAuthentication(AuthenticationContextInterface $context): bool;
 
+    /**
+     * Determine whether this Provider needs to be prepared (if the prepareAuthentication method needs to be called).
+     */
+    public function needsPreparation(): bool;
+
     /**
      * Do all steps necessary to prepare authentication, e.g. generate & send a code.
```

Comment:
To make the new behavior more clear:

```suggestion
     * Do all steps necessary to prepare authentication, e.g. generate & send a code. Will only be called when needsPreparation() returns true.
```